### PR TITLE
JS: improve support for route handlers in containers

### PR DESCRIPTION
**Pull request by esbena**
This is a proposal a step towards recognizing additional route handlers at our route setups.

A common theme in many advanced server applications is that route handlers are defined in one corner of the application, and then registered in route setups in a different corner using an intermediary container for transport. This intermediary container strains our current backwards type tracking for route handlers (See some of the many test cases added in this PR).

While the purest way of supporting this pattern is to extend our general type tracking, I think this PR is a sweet spot step that can leverage our existing route handler heuristics.

For team-context, this PR is safer than <https://github.com/github/codeql-javascript-team/issues/57>, and inspired by <https://github.com/github/codeql-javascript-team/issues/52>.

I am not very opposed to dropping this PR completely, but I think we should at least have a good discussion (somewhere) about the likely ways forward.

The performance seems unchanged on a preliminary [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/bulky-route-handler-registration_1581045305786).

### CODE REVIEW

**Comment by esbena** [comment]
(I have snuck in some additional tests I had around.)

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 1]
```diff
@@ -475,4 +476,118 @@ module HTTP {
...
+   */
+  class RouteHandlerCandidateContainer extends DataFlow::Node {
+    RouteHandlerCandidateContainer::Range self;
+
+    RouteHandlerCandidateContainer() { this = self }
+
+    /**
+     * Gets the route handler in this container that is accessed at `access`.
+     */
+    DataFlow::SourceNode getRouteHandler(DataFlow::SourceNode access) {
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]: All implementations of this method return a `RouteHandlerCandidate`, so shouldn't the type reflect this? 
(Then you can remove the two `result instanceof RouteHandlerCandidate`)

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 2]
```diff
@@ -475,4 +476,118 @@ module HTTP {
...
+          or
+          exists(EnumeratedPropName prop | access = prop.getASourceProp())
+          or
+          read = DataFlow::lvalueNode(any(ForOfStmt stmt).getLValue())
+        )
+      }
+    }
+
+    /**
+     * A map that contains one or more route potential handlers.
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]:
```suggestion
     * A `Map` object that contains one or more route potential handlers.
```
I would like to somehow have an emphasis that this is an instance of `new Map()`, and not just any map implementation.

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 3]
```diff
@@ -475,4 +476,118 @@ module HTTP {
...
+        |
+          exists(DataFlow::MethodCallNode get, DataFlow::Node getKey |
+            get = access and
+            getKey = get.getArgument(0) and
+            ref(this).getAMethodCall("get") = get
+          |
+            exists(string name |
+              getKey.mayHaveStringValue(name) and
+              setKey.mayHaveStringValue(name)
+            )
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]: Could this be too restrictive? 

I have no idea if it happens in real projects, but I could imagine some code where the keys are dynamically computed (or the keys are symbols?). 
Could it be an idea to change the logic to: `(if setter key is known) implies (setter key = getter key)`?

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 4]
```diff
@@ -475,4 +476,118 @@ module HTTP {
...
+          )
+          or
+          exists(DataFlow::MethodCallNode forEach |
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]: What about `for ([key, value] of new Map()) {..}` / `for (entry of new Map()) {..}`?

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll (hunk 1 above)
Comment (esbena) [reply javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]: The intention was to hide the fact that `RouteHandlerCandidate` was used under the hood, but I guess the `RouteHandlerCandidateContainer` name is a spoiler in that regard :). I will leave it as is, and see if we end up changing the `RouteHandlerCandidateContainer` name instead due to some other discussion...

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll (hunk 2 above)
Comment (esbena) [reply javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]: I am hoping to improve this with https://github.com/github/codeql-javascript-team/issues/56 later.

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll (hunk 3 above)
Comment (esbena) [reply javascript/ql/src/semmle/javascript/frameworks/HTTP.qll]: For both `ContainerObject` and `ContainerMap`, I have chosen to exclude keys that we have no idea about the value of. We should definitely try to loosen up this restriction later, but that requires a separate large evaluation IMO.

**Comment by erik-krogh** [comment]
Now that https://github.com/Semmle/ql/pull/3169 is merged, I used the Collection steps from that PR to make a slight generalization of your `ContainerMap`. 

Commit: https://github.com/Semmle/ql/commit/15acc39612bb1accd78ea81eba066b92e5b45feb  

You can merge that into this PR if you want.

**Comment by esbena** [comment]
Thanks. I will be able to do a query console run once the dist-upgrade completes.

**Comment by esbena** [comment]
This seems promising, but still a bit underwhelming. [This query console run](https://lgtm.com/query/390812634926874877) results in [this spreadsheet](https://docs.google.com/spreadsheets/d/1gQE4AWKZ2EUI_hPPwzdAWTbUmQ096R-ikLKD0rbuDaQ/edit?usp=sharing) which shows that we now identify 398 new `Express::RouteHandlers` in 35 of 120k projects.
I'll check the performance, and verify that the results above are TPs...

**Comment by esbena** [comment]
The results look good. Just waiting for the performance now.

The most common new route handler we identify is the deeply nested one:

```js
app.get('/api/component', routes.componentRegistry.components);
```

One particularly nice new route handler to see in practice is this one:
```js
var app = makeapp();
Object.keys(routes).forEach(function (route) {
  app.get(route, routes[route]);
})


```

**Comment by asgerf** [comment]
Thanks for getting these numbers. 👍 Happy to see this moving forward again; let's hope performance works out.

**Comment by esbena** [comment]
The [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/bulky-route-handler-registration_1591373517355) is benign.

I did try out https://github.com/github/codeql/commit/15acc39612bb1accd78ea81eba066b92e5b45feb, but had simplify it (see https://github.com/github/codeql/pull/2801/commits/f618d430e7274de8ddd5c0140e7009f9856a2668) since it [did not scale](https://git.semmle.com/esben/dist-compare-reports/tree/js/bulky-route-handler-registration_1591267806250) as it also type tracked route handlers once they exited their container, and *that* is the responsibility of `RouteSetup::getARouteHandler`.

**Comment by esbena** [comment]
Tests are green

**Comment by esbena** [comment]
Ping. This is still ready for merge IMO.

**Pull request merged**